Sources/RemoteMessaging/RemoteMessagingProcessing.swift

```diff
@@ -0,0 +1,101 @@
+}
+
+/**
+ * This protocol defines API for Remote Messaging client in the app.
+ */
+public protocol RemoteMessagingProcessing {
+    /// Defines endpoint URL where the config file is available.
+    var endpoint: URL { get }
+
+    /// This holds the fetcher that downloads the config file from the server.
+    var configurationFetcher: RemoteMessagingConfigFetching { get }
+
+    /// This holds the config matcher provider that updates the config matcher before the config is evaluated.
+    var configMatcherProvider: RemoteMessagingConfigMatcherProviding { get }
+
+    /// Provides feature flag support for RMF.
+    var remoteMessagingAvailabilityProvider: RemoteMessagingAvailabilityProviding { get }
+
+    /**
+     * This function returns a config processor.
+     *
+     * Config processor performs evaluation of the JSON config file against the matcher containing
+     * app-specific, device-specific and user-specific matched attributes. Default implementation is provided.
+     */
+    func configProcessor(for configMatcher: RemoteMessagingConfigMatcher) -> RemoteMessagingConfigProcessing
+
+    /**
+     * This is the entry point to RMF from the client app.
+     *
+     * This function fetches the config, updates config matcher, evaluates the config against the matcher
+     * and stores the result as needed. Client apps should call this function in order to refresh remote messages.
+     * When messages are updated, `RemoteMessagingStore.Notifications.remoteMessagesDidChange` notification is posted.
+     * Default implementation is provided.
+     */
+    func fetchAndProcess(using store: RemoteMessagingStoring) async throws
+}
+
+public extension RemoteMessagingProcessing {
+
+    func configProcessor(for configMatcher: RemoteMessagingConfigMatcher) -> RemoteMessagingConfigProcessing {
+        RemoteMessagingConfigProcessor(remoteMessagingConfigMatcher: configMatcher)
+    }
+
+    func fetchAndProcess(using store: RemoteMessagingStoring) async throws {
+        guard remoteMessagingAvailabilityProvider.isRemoteMessagingAvailable else {
+            os_log("Remote messaging feature flag is disabled, skipping fetching messages", log: .remoteMessaging, type: .debug)
+            return
+        }
+        do {
+            let statusResponse = try await configurationFetcher.fetchRemoteMessagingConfig()
+
+            os_log("Successfully fetched remote messages", log: .remoteMessaging, type: .debug)
+
+            let remoteMessagingConfigMatcher = await configMatcherProvider.refreshConfigMatcher(using: store)
+
+            let processor = configProcessor(for: remoteMessagingConfigMatcher)
+            let config = store.fetchRemoteMessagingConfig()
+
+            if let processorResult = processor.process(jsonRemoteMessagingConfig: statusResponse, currentConfig: config) {
+                store.saveProcessedResult(processorResult)
+            }
```

Comment:
perhaps this should be guard that throws some processing failed error? otherwise we're not getting any debug info what happened

Reply:
This method calls a bunch of Store private methods inside, each of which can throw an error that is handled by sending a pixel (via `EventMapping`) and printing a console log. So I think it's acceptable. I agree that it's not ideal though.

Also, this function's error is silenced at the call site (with `try?` – again probably acceptable in this specific case because it effectively silences only network and JSON parsing errors). I may propose an update to this piece of code but it has been a status quo for past 2 years so I'd rather handle it separately.

Reply:
don't bother, makes sense to me